binder/src/androidTest/java/io/grpc/binder/internal/BinderClientTransportTest.java

```diff
@@ -334,41 +321,71 @@ public void transportInUse(boolean inUse) {
 
   private static final class TestStreamListener implements ClientStreamListener {
 
-    public StreamListener.MessageProducer messageProducer;
     public boolean ready;
     public Metadata headers;
-    @Nullable public Status closedStatus;
+
+    @GuardedBy("this")
+    private final Deque<MessageProducer> messageProducers = new ArrayDeque<>();
+
+    @GuardedBy("this")
+    @Nullable
+    private Status closedStatus;
 
     @Override
-    public void messagesAvailable(StreamListener.MessageProducer messageProducer) {
-      this.messageProducer = messageProducer;
+    public synchronized void messagesAvailable(StreamListener.MessageProducer messageProducer) {
+      messageProducers.add(messageProducer);
+      notifyAll();
     }
 
-    public synchronized void awaitMessages() {
-      while (messageProducer == null) {
-        try {
-          wait(100);
-        } catch (InterruptedException inte) {
-          throw new AssertionError("Interrupted waiting for messages");
+    /** Blocks until at least one MessageProducer has been provided for reading. */
+    public synchronized void awaitMessages() throws InterruptedException {
+      while (messageProducers.isEmpty()) {
+        wait();
+      }
+    }
+
+    /** Blocks until {@code n} messages can be produced (and discarded). */
+    public synchronized void readAndDiscardMessages(int n)
+        throws InterruptedException, IOException {
+      while (n > 0) {
+        while (closedStatus == null && messageProducers.isEmpty()) {
+          wait();
+        }
+        if (closedStatus != null) {
+          throw closedStatus.withDescription("premature close").asRuntimeException();
+        }
+        try (InputStream message = messageProducers.peek().next()) {
+          if (message == null) {
```

Comment:
You aren't iterating the messages within a messageProducer, just the messageProducers themselves.

Reply:
assuming this is just a duplicate of the comment below ... ?

Reply:
Yes, it is the same situation, but I only see the break replaced at line 360, not on line 385.

Reply:
done thank you